### PR TITLE
Added optional encryption block for azurerm_recovery_services_vault

### PR DESCRIPTION
**Pull request by shanoor**
## PR Checklist

---

- [x] I have added example(s) inside the [./examples/] folder
- [x] I have added the example(s) to the integration test list for [normal (~30 minutes)](./workflows/standalone-scenarios.json) or [long runner >30 minutes](./workflows/standalone-scenarios-longrunners.json)
- [x] I have checked the [coding conventions as per the wiki](https://github.com/aztfmod/terraform-azurerm-caf/wiki)
- [x] I have checked to ensure there aren't other open Pull Requests for the same update/change?

## Description

This allows to use a custom encryption key on a recovery vault by using the `encryption` block.

## Does this introduce a breaking change

- [ ] YES
- [x] NO

## Testing

```
recovery_vaults = {
  asr1 = {
    name               = "vault_re1"
    resource_group_key = "primary"

    region = "region1"

    replication_policies = {
      repl1 = {
        name               = "policy1"
        resource_group_key = "primary"

        recovery_point_retention_in_minutes                  = 24 * 60
        application_consistent_snapshot_frequency_in_minutes = 4 * 60
      }
    }


    backup_policies = {
      vms = {
        policy1 = {
          name                           = "VMBackupPolicy1"
          vault_key                      = "asr1"
          rg_key                         = "primary"
          timezone                       = "UTC"
          instant_restore_retention_days = 5
          backup = {
            frequency = "Daily"
            time      = "23:00"
            #if not desired daily, can pick weekdays as below:
            #weekdays  = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
          }
          retention_daily = {
            count = 10
          }
        }
      }
    }

    encryption = {
      key_id                            = "https://myvault001.vault.azure.net/keys/asr-cmk/"
      infrastructure_encryption_enabled = true
    }
  }
}
```

### CODE REVIEW

File: examples/recovery_vault/106-asr-with-custom-encryption-key/configuration.tfvars [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            weekdays = ["Sunday"]
+            weeks    = ["Last"]
+            months   = ["January"]
+          }
+        }
+      }
+    }
+
+    encryption = {
+      key_id                            = "https://myvault001.vault.azure.net/keys/asr-cmk/"
```
Comment (arnaudlh) [inline examples/recovery_vault/106-asr-with-custom-encryption-key/configuration.tfvars:99]: Would need the option to link it with the caf module object-model (key vault key and remote key)

File: modules/recovery_vault/recovery_vault.tf [hunk 2]
```diff
@@ -24,4 +24,13 @@ resource "azurerm_recovery_services_vault" "asr" {
     type = "SystemAssigned"
   }
 
+  dynamic "encryption" {
+    for_each = can(var.settings.encryption) ? [1] : []
+
+    content {
+      key_id                            = var.settings.encryption.key_id
```
Comment (arnaudlh) [inline modules/recovery_vault/recovery_vault.tf:31]: Would need the option to link it with the caf module object-model (key vault key and remote key). Send the key vault keys and find the index from here?

**Review by arnaudlh — CHANGES_REQUESTED** [review]
Looks good @shanoor, but on top the key_id in plain text, would be nice to have the reference to the object model for keys created with the module.

File: modules/recovery_vault/recovery_vault.tf (hunk 2 above)
Comment (shanoor) [reply modules/recovery_vault/recovery_vault.tf:31]: I tried to do that before submitting the PR but there is an huge circular reference error. I spent few hours trying to figure out how to solve it but to no avail.

File: examples/recovery_vault/106-asr-with-custom-encryption-key/configuration.tfvars (hunk 1 above)
Comment (shanoor) [reply examples/recovery_vault/106-asr-with-custom-encryption-key/configuration.tfvars:99]: I tried to do that before submitting the PR but there is an huge circular reference error. I spent few hours trying to figure out how to solve it but to no avail.

File: modules/recovery_vault/recovery_vault.tf (hunk 2 above)
Comment (LaurentLesle) [reply modules/recovery_vault/recovery_vault.tf:31]: ok let's progress one step at a time. At least this PR add support for CMK using key resource id. We can figure-out later for the keyvault support. One way to break the circular ref is to only reference a remote keyvault. In real life we see customers centralising the same key for their keyvaults. So remote keyvault would be my preference here.
@shanoor  Let me know if you can try that scenario, if not I merge this PR.

Comment (shanoor) [reply modules/recovery_vault/recovery_vault.tf:31]: I tried with

```tf
encryption_key_id = try(each.value.encryption.key_id, local.combined_objects_keyvault_keys[each.value.encryption.lz_key][each.value.encryption.key_key], null)
```

A plan without even referencing any remote key fails with the circular reference error:

> Error: Cycle: module.solution.module.keyvault_access_policies.module.diagnostic_storage_accounts.azurerm_key_vault_access_policy.policy, module.solution.module.keyvault_access_policies.module.diagnostic_storage_accounts.var.tenant_id (expand), module.solution.module.keyvault_access_policies.module.diagnostic_storage_accounts.var.object_id (expand), module.solution.module.keyvault_access_policies.module.diagnostic_storage_accounts (close), module.solution.module.keyvault_access_policies.module.mssql_managed_instances_secondary.var.object_id (expand), module.solution.module.keyvault_access_policies.module.mssql_managed_instances_secondary.azurerm_key_vault_access_policy.policy, module.solution.module.keyvault_access_policies.module.mssql_managed_instances_secondary (close), module.solution.module.keyvault_access_policies.module.azuread_service_principals.var.object_id (expand), module.solution.module.keyvault_access_policies.module.azuread_service_principals.azurerm_key_vault_access_policy.policy, module.solution.module.keyvault_access_policies.module.azuread_service_principals.var.tenant_id (expand), module.solution.module.keyvault_access_policies.module.azuread_service_principals (close), module.solution.module.keyvault_access_policies.module.storage_accounts.var.tenant_id (expand), module.solution.module.keyvault_access_policies.module.storage_accounts.var.object_id (expand), module.solution.module.keyvault_access_policies.module.storage_accounts.azurerm_key_vault_access_policy.policy, module.solution.module.keyvault_access_policies.module.storage_accounts (close), module.solution.module.keyvault_access_policies.module.mssql_managed_instance.var.object_id (expand), module.solution.module.keyvault_access_policies.module.mssql_managed_instance.azurerm_key_vault_access_policy.policy, module.solution.module.keyvault_access_policies.module.mssql_managed_instance (close), module.solution.module.storage_accounts.output.backup_container_id (expand), module.solution.module.storage_accounts.module.file_share.azurerm_backup_protected_file_share.fs_backup, module.solution.module.storage_accounts.module.file_share.output.name (expand), module.solution.module.storage_accounts.module.file_share.output.id (expand), module.solution.module.storage_accounts.module.file_share.var.storage_account_id (expand), module.solution.module.storage_accounts.module.file_share.output.resource_manager_id (expand), module.solution.module.storage_accounts.module.file_share.output.file_share_directories (expand), module.solution.module.storage_accounts.module.file_share.var.resource_group_name (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_file.var.share_id (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_file.var.settings (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_file.azurerm_storage_share_file.share_file, module.solution.module.storage_accounts.module.file_share.module.file_share_directory.output.name (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_directory.var.storage_account_name (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_directory.var.share_name (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_directory.azurerm_storage_share_directory.share_directory, module.solution.module.storage_accounts.module.file_share.module.file_share_directory.output.id (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_directory (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_directory.var.settings (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_directory (close), module.solution.module.storage_accounts.module.file_share.module.file_share_file (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_file.output.id (expand), module.solution.module.storage_accounts.module.file_share.module.file_share_file (close), module.solution.module.storage_accounts.module.file_share.var.recovery_vault (expand), module.solution.module.storage_accounts.module.file_share (close), module.solution.module.storage_accounts.module.file_share.var.storage_account_name (expand), module.solution.module.storage_accounts.module.file_share.var.settings (expand), module.solution.module.storage_accounts.module.file_share.azurerm_storage_share.fs, module.solution.module.storage_accounts.module.file_share (expand), module.solution.module.storage_accounts.module.file_share.output.url (expand), module.solution.module.storage_accounts.output.file_share (expand), module.solution.module.recovery_vaults.output.backup_policies (expand), module.solution.module.recovery_vaults.module.private_endpoint.output.id (expand), module.solution.module.recovery_vaults.module.private_endpoint.output.private_dns_zone_group (expand), module.solution.module.recovery_vaults.module.private_endpoint.azurerm_private_endpoint.pep, module.solution.module.recovery_vaults.module.private_endpoint.output.private_dns_zone_configs (expand), module.solution.module.recovery_vaults.module.private_endpoint.var.resource_id (expand), module.solution.module.recovery_vaults.module.private_endpoint (close), module.solution.module.recovery_vaults.azurerm_backup_policy_file_share.fs, module.solution.module.recovery_vaults.azurerm_site_recovery_network_mapping.network_mapping, module.solution.module.recovery_vaults.module.diagnostics.azurerm_monitor_diagnostic_setting.diagnostics, module.solution.module.recovery_vaults.module.diagnostics.var.resource_id (expand), module.solution.module.recovery_vaults.module.diagnostics (close), module.solution.module.recovery_vaults.azurerm_backup_policy_vm.vm, module.solution.module.recovery_vaults.output.rbac_id (expand), module.solution.module.recovery_vaults.azurerm_site_recovery_protection_container.protection_container, module.solution.module.recovery_vaults.azurerm_site_recovery_fabric.recovery_fabric, module.solution.module.recovery_vaults.azurerm_site_recovery_protection_container_mapping.container-mapping, module.solution.module.recovery_vaults (close), module.solution.module.recovery_vaults.time_sleep.delay_create, module.solution.module.recovery_vaults.azurerm_site_recovery_replication_policy.policy, module.solution.module.recovery_vaults.output.replication_policies (expand), module.solution.module.recovery_vaults.output.id (expand), module.solution.module.keyvault_keys.output.public_key_openssh (expand), module.solution.module.keyvault_keys.output.public_key_pem (expand), module.solution.module.keyvault_keys.output.versionless_id (expand), module.solution.module.keyvault_keys.output.name (expand), module.solution.module.keyvault_keys.output.id (expand), module.solution.module.keyvault_keys.var.keyvault (expand), module.solution.module.keyvault_keys.var.global_settings (expand), module.solution.module.keyvault_keys.local.base_tags (expand), module.solution.module.keyvault_keys.local.tags (expand), module.solution.module.keyvault_keys.var.settings (expand), module.solution.module.keyvault_keys (expand), module.solution.module.keyvault_keys.azurerm_key_vault_key.key, module.solution.module.keyvault_keys (close), module.solution.local.combined_objects_keyvault_keys (expand), module.solution.module.recovery_vaults.var.encryption_key_id (expand), module.solution.module.recovery_vaults.azurerm_recovery_services_vault.asr, module.solution.module.recovery_vaults.output.name (expand), module.solution.local.combined_objects_recovery_vaults (expand), module.solution.module.storage_accounts.var.recovery_vaults (expand), module.solution.module.storage_accounts.local.recovery_vault (expand), module.solution.module.storage_accounts.azurerm_backup_container_storage_account.container, module.solution.module.storage_accounts (close), module.solution.local.combined_objects_storage_accounts (expand), module.solution.module.keyvault_access_policies.var.resources (expand), module.solution.module.keyvault_access_policies.module.managed_identity.var.object_id (expand), module.solution.module.keyvault_access_policies.module.managed_identity.azurerm_key_vault_access_policy.policy, module.solution.module.keyvault_access_policies.module.managed_identity (close), module.solution.module.keyvault_access_policies (close)
```

**Comment by arnaudlh** [comment]
thanks for the PR @shanoor, I will close this one as #1697 seems more modern. Please add examples to #1697 and we should be good!

**Pull request closed without merging**